### ISSUE TITLE
On Fedora 24 cannot download EDG binary

### ISSUE THREAD

**Issue by ian-bertolacci** [rose-compiler/rose#23]
During compilation, attempts to download the the EDG binary from the url http://rosecompiler.org/edg_binaries/roseBinaryEDG-4-9-x86_64-pc-linux-gnu-gnu-6.3-4.a4817f0e9317c375b92aeb67ff7666f46f898452.tar.gz
However, that url points, essentially, to the rosecompiler.org home page.

wget doesn't particularly care, downloads the html contents as roseBinaryEDG-4-9-x86_64-pc-linux-gnu-gnu-6.3-4.a4817f0e9317c375b92aeb67ff7666f46f898452.tar.gz.
gzip/tar then comes along and tries to unzip the file resulting in the error: `gzip: stdin: not in gzip format`

On Fedora release 24 (Twenty Four) (Linux hostname 4.9.17-100.fc24.x86_64 #1 SMP Wed Mar 22 15:49:46 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux).
Using the rose-compiler/rose repository at either the v0.9.9.0 release tag or the most recent commit.

**Comment by thiagotei** [comment]
You should read this: [ Installing EDG](http://rosecompiler.org/ROSE_HTML_Reference/group__installing__edg.html)
They mention about this kind of failure there.

**Comment by shivansh** [comment]
After configuring the rose environment (before running `make`), I was trying to install EDG following [this link](http://rosecompiler.org/ROSE_HTML_Reference/group__installing__edg.html).
On running `make show-edg-url` from `src/frontend/CxxFrontend` in the build tree, I got this URL - http://rosecompiler.org/edg_binaries/roseBinaryEDG-4-9-x86_64-pc-linux-gnu-gnu-6.3-4.a4817f0e9317c375b92aeb67ff7666f46f898452.tar.gz/
However, I don't think this is the correct URL as it points to a webpage rather than a tar file.
```
▶ wget http://www.rosecompiler.org/edg_binaries/roseBinaryEDG-4-9-x86_64-pc-linux-gnu-gnu-6.3-4.a4817f0e9317c375b92aeb67ff7666f46f898452.tar.gz

▶ file roseBinaryEDG-4-9-x86_64-pc-linux-gnu-gnu-6.3-4.a4817f0e9317c375b92aeb67ff7666f46f898452.tar.gz
roseBinaryEDG-4-9-x86_64-pc-linux-gnu-gnu-6.3-4.a4817f0e9317c375b92aeb67ff7666f46f898452.tar.gz: HTML document, ASCII text, with very long lines, with CRLF, LF line terminators
```

**Comment by shivansh** [comment]
I got this URL on running `make show-edg-url` - http://www.rosecompiler.org/edg_binaries/roseBinaryEDG-4-9-x86_64-pc-linux-gnu-gnu-6.3-4.a4817f0e9317c375b92aeb67ff7666f46f898452.tar.gz ; which is essentially a web-page (as pointed above).
The page [Installing EDG](http://rosecompiler.org/ROSE_HTML_Reference/group__installing__edg.html) does mention that it may be possible that a wrong URL can be generated. However, what can be done in this case ?

**Comment by laraaydin** [comment]
This is also a problem on Ubuntu. A false _edg signature_ is generated, and the link does not properly handle this problem, thus the build process downloads an html page instead of a tar archive. However, you can manipulate your _DownloadEDG.cmake_ (like in here: https://github.com/rose-compiler/rose-develop/blob/master/cmake/DownloadEDG.cmake) to look for a specific edg signature, which you can find one here: http://rosecompiler.org/edg_binaries/ . I still haven't tried this approach, but it feels like it will work well! :)

**Comment by ian-bertolacci** [comment]
I have rose building pretty consistantly on ubuntu 16.04 with gcc5/g++5. Maybe something about your particular setup?

I will keep the cmake trick in mind.

**Comment by laraaydin** [comment]
I will try that right now. I was using gcc4.9/g++4.9. Could you also tell me your boost version? Thank you! :)

**Comment by ian-bertolacci** [comment]
I use boost 1.61
I *feel* like gcc/g++ 4.9 should work, but I dunno. 

We have a docker image based on ubuntu 16.04 that we can build rose in.
Here is what we install on it, using apt (some of this is unnecessary):
mercurial git subversion wget csh tar unzip make autoconf automake cmake libtool libtool-bin perl python3 gcc-5 g++-5 default-jdk default-jre flex bison libarchive-dev libbz2-dev libgmp-dev libgmp10 libgmpxx4ldbl zlib1g zlib1g-dev lsb-release ghostscript

**Comment by laraaydin** [comment]
I am also trying to build a docker image based on ubuntu that also runs some scripts to install the basic required dependencies for ROSE, and the problem with it is that it cannot produce valid edg signatures for some reason. I tried to only change gcc4.9/g++4.9 to gcc5/g++5 and this time I got these errors:
```
configure: error: in `/build-rose':
configure: error: C compiler cannot create executables
```
I will keep trying though, and thank you so much for your help!

**Comment by shivansh** [comment]
The problems which I pointer out earlier were on my Ubuntu 17.04 machine. I just had a successful build on Ubuntu 16.04 !

**Comment by ian-bertolacci** [comment]
@teamblubee my understanding is that the EDG parser is closed source, which is why they have this complicated binay distribution system.

**Comment by teamblubee** [comment]
would it be possible to use something like Clang as a frontend for c/c++ code?

**Comment by ian-bertolacci** [comment]
Possible? Sure. But it's very unlikely that they will add that kind of support, and I'm sure reengineering the computer to do so would be very difficult.

**Comment by justintoo** [comment]
Here's a summary of tips:

1. Clang support has been a work in progress :-) `ROSE/configure --enable-clang-frontend`
2. Official docker images are a current work in progress - rolling out over the next month
3. Binary distributions are also a work in progress with a few already available for RHEL7 (and compatible platforms): http://rosecompiler.org/rose_binaries/. Official documentation has `v0.9.9.171` and will be regularly updated in the near future: http://rosecompiler.org/ROSE_HTML_Reference/group__installation.html.
4. Here is some documentation on EDG tarballs: http://rosecompiler.org/ROSE_HTML_Reference/group__installing__edg.html. We do not support every OS (yet) and EDG source code is only available to those with a commercial/research license.

Thanks for your patience as we continue to ramp up support across these topics over the next few months.

**Comment by teamblubee** [comment]
Hi @justintoo 

I did read all the documentation on EDG and looked at the official tested configurations, I also reached out to the EDG team but I was informed that I'd have to take up the request with ROSE devs about getting access to source code to build custom binaries, not only that but the EDG representative said they did not have nor use FreeBSD so I think it might be a simpler solution to work with the Clang frontend.

I can imagine that at the time ROSE was initially being developed EDG could've been one of the most advanced C/C++ compiler frontend but LLVM's Clang has come a long way in that regard since then.

Are you saying it's possible to build ROSE without EDG, especially on a platform where there's no binaries available?

Or is it like some of the other components that are required to be built even if you have no plans of using that particular frontend?

**Comment by ian-bertolacci** [comment]
@teamblubee it might be useful to build in a docker container. I think it makes development slightly more pained, but it may work for you. I have a rather complicated docker image and scripted build setup for a project I am working on. If you'd like I can try and make it Rose centric.

**Comment by milahu** [comment]
still an issue

seems like the edg binaries are not provided any more?

directory listing is disabled for http://rosecompiler.org/edg_binaries/
so i tried to brute-force the edg url for the new signature format `5.x.x.x.x`
example signarure: `5.0.11.82.3` where `0.11.82.3` is the [EDG_VERSION](https://github.com/rose-compiler/rose/blob/weekly/src/frontend/CxxFrontend/EDG_VERSION)
with url format `http://rosecompiler.org/edg_binaries/roseBinaryEDG-$v.tar.gz`
and `$v` values in [edg_versions_not_found.txt](https://github.com/rose-compiler/rose/files/8761224/edg_versions_not_found.txt)
... but no luck

> Clang support has been a work in progress :-) `ROSE/configure --enable-clang-frontend`

sounds better than chasing edg binaries
im adding the `enable-clang-frontend` option for cmake in #195

**Comment by pinnown** [comment]
@milahu EDG binaries are still being generated, you can use `make show-edg-url` to show the URL. My assumption is you are using a compiler that we don't release a binary for, currently we release a binary for GCC 6 through 10.

**Comment by milahu** [comment]
> My assumption is you are using a compiler that we don't release a binary for

i have tried ...

> and `$v` values in [edg_versions_not_found.txt](https://github.com/rose-compiler/rose/files/8761224/edg_versions_not_found.txt)

for example

```
curl -sI http://rosecompiler.org/edg_binaries/roseBinaryEDG-5-0-x86_64-pc-linux-gnu-8.0-5.0.11.80.1.tar.gz | head -n1
HTTP/1.1 404 Not Found

curl -sI http://rosecompiler.org/edg_binaries/roseBinaryEDG-5-0-x86_64-pc-linux-gnu-gnu-8.0-5.0.11.82.3.tar.gz | head -n1
HTTP/1.1 404 Not Found
```

can you please give some examples of working urls?

**Comment by pinnown** [comment]
I didn't notice you were using the wrong EDG URL in the original post. 
```
wget http://edg-binaries.rosecompiler.org/roseBinaryEDG-5-0-x86_64-pc-linux-gnu-gnu-8-5.0.11.80.1.tar.gz
```

The CMake build system was not maintained and is currently being refactored. I would recomend using the autoconf build system. https://github.com/rose-compiler/rose/wiki/Install-Rose-From-Source

**Comment by milahu** [comment]
> The CMake build system was not maintained and is currently being refactored.

aah ... i posted some fixes in #195 

for me (on nixos) both are broken (autoconf cannot find boost lib), and fixing cmake makes more sense
